### ISSUE TITLE
Gerrit source step doesn't work well with multiple codebases

### ISSUE THREAD

**Issue by willlovett-arm** [buildbot/buildbot#3460]
I have a project that consists of multiple codebases, usually with the same branch names, within a gerrit instance.

My GerritChangeSource monitors them and performs pre-commit testing on the project (on patchset-created events) by checking out each of these codebases using a Gerrit step.

However, Gerrit.build.getProperty("event.patchSet.ref") is not currently indexed by codebase, so every Gerrit step attempts to checkout the change ref, which fails on all but the modified codebase.

I currently work around this in my buildbot config by subclassing Gerrit:

```
from buildbot.steps.source.gerrit import Gerrit

class MultiCodebaseGerrit(Gerrit):
    def __init__(self, **kwargs):
        Gerrit.__init__(self, **kwargs)

    def startVC(self, branch, revision, patch):
        if self.sourcestamp.codebase == self.codebase:
            super(Gerrit, self).startVC(branch, revision, patch)
        else:
            super(Git, self).startVC(branch, revision, patch)
```

1. Am I missing some easier way of doing this?, and if not
2. Is the patch below roughly the right way of fixing it?  If so, I'll create a pull request
```
diff --git a/master/buildbot/steps/source/gerrit.py b/master/buildbot/steps/source/gerrit.py
index 979f77ce0..4c70ab184 100644
--- a/master/buildbot/steps/source/gerrit.py
+++ b/master/buildbot/steps/source/gerrit.py
@@ -26,7 +26,11 @@ class Gerrit(Git):

     def startVC(self, branch, revision, patch):
         gerrit_branch = None
-        if self.build.hasProperty("event.patchSet.ref"):
+        if self.sourcestamp.codebase != self.codebase:
+            # The change is from a different codebase, so just check out the
+            # head of the given branch
+            pass
+        elif self.build.hasProperty("event.patchSet.ref"):
             gerrit_branch = self.build.getProperty("event.patchSet.ref")
             self.updateSourceProperty("gerrit_branch", gerrit_branch)
         else:
```

Thanks!

Will Lovett

**Comment by tardyp** [comment]
The patch looks good. Please send a PR, and don't forget to update the unit tests so that we don't forget about this use case.

docs.buildbot.net/latest/developer/tests.html

**Comment by p12tic** [comment]
I think this bug was fixed in #3645. Closing
